### PR TITLE
[node],[mongodb] Readable stream interface, Closes #13782, Closes #13678

### LINKED ISSUE #13678: @types/mongodb Incorrect definition in @types/mongodb
It looks like the pipe method of mongodb cursor has incorrect definition, this makes it very difficult to use in typescript projects.

This is the current line:

```
//http://mongodb.github.io/node-mongodb-native/2.1/api/Cursor.html#pipe
    pipe(destination: Writable, options?: Object): void;
```

But the return type should be a destination stream, not void.
This is how the ReadableStream.pipe is typed in node's own definition file
```
pipe<T extends WritableStream>(destination: T, options?: { end?: boolean; }): T;
```

With the current definitions it's difficult to use .pipe method on mongo cursor without having typescript compiler errors. To be most specific, it's not possible to chain calls to .pipe now because of return type void

For example this should and actually does return a new stream, so the result of this call is a ReadableStream but because of this incorrect return type **void** attempting to chain another pipe call generates tsc compiler error

```
// here the res is node.js response object http.ServerResponse, which is a Writable stream
cursor.pipe(JSONStream.stringify()).pipe(res)
```

### LINKED ISSUE #13782: @types/node incorrect API/differs from official
## Origin of the issue
The return types of `ReadableStream.setEncoding()` method according to
[nodejs v0.12 API](https://nodejs.org/dist/v0.12.0/docs/api/stream.html#stream_readable_setencoding_encoding) is set to `this` 
[node.js 6x API ](https://nodejs.org/dist/latest-v6.x/docs/api/stream.html#stream_readable_setencoding_encoding) is set to `this`<br> and [node/index.d.ts](https://github.com/hinell/DefinitelyTyped/blob/master/node/index.d.ts#L284) is set to`void` ?!

These are also have return type defined incorrectly:
 ```ts
readable.pause()
readable.resume()
readable.unpipe()
```
All of them expected to return `this` instead of `void/T` as defined by latest node.d.ts.

Obviously that discrepancy requires fix. Essentially there is also need to prepare a list of modules that are dependent on ReadableStream interface (inherit/extends it) to fix them up and avoid any errors like this one #13678. 

## Blank
- [X] I tried using the latest `mongodb/index.d.ts` file with [node.d.ts](https://github.com/types/env-node/tree/master/6.0) provided by [typings](https://github.com/typings/typings)  in this repo and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [  ] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [X] I want to talk about `node/index.d.ts`
  - The authors of that type definition are 
@andy-ms
@RyanCavanaugh
@minestarks
@paulvanbrenk
@sandersn
@maghis
@zhengbli
@yuit
@zjiekai
@yortus
@xiemaisi
@rogierschouten
@rakatyal
@mhegazy
@iainmcgin
@horiuchi
@HerringtonDarkholme
@jefff
@Flarna
@icopp
@a8775

### PR DESCRIPTION
**Pull request by hinell**
### [node]
The return types of next`Readable` interface's  methods were changed onto
`this` type according to their officially declared api: [setEncoding()](https://github.com/nodejs/node/blob/v6.9.4/lib/_stream_readable.js#L221) [pause()](https://nodejs.org/dist/latest-v6.x/docs/api/stream.html#stream_readable_pause) [resume()](https://nodejs.org/dist/latest-v6.x/docs/api/stream.html#stream_readable_resume) [unpipe()](https://github.com/nodejs/node/blob/v6.9.4/lib/_stream_readable.js#L681)
`ConsoleConstructor` and its tests were both added

### [mongodb]
'Readable'-dependent types were reconciled (removed) with changed `node/index.d.ts` api

## Issues to close
#13782
#13678

## Template
Please fill in this template.

- [X] Make your PR against the `master` branch.
- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped#make-a-pull-request).
- [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped#common-mistakes).
- [X] Run `tsc` without errors.

Select one of these and delete the others:
If changing an existing definition:
- [X] Provide a URL to documentation or source code which provides context for the suggested changes: see issue here #13782

### CODE REVIEW

**Comment by CaselIT** [comment]
Hi, could you remove the remove the methods in mongodb, since they have been mistakenly been added there? We should use the one on the parent class.

Thanks

**Comment by hinell** [comment]
>could you remove the remove the methods in mongodb
@CaselIT What the methods?

File: mongodb/index.d.ts [hunk 1]
```diff
@@ -1185,7 +1185,7 @@ export interface Cursor extends Readable {
     // http://mongodb.github.io/node-mongodb-native/2.1/api/Cursor.html#setCursorOption
     setCursorOption(field: string, value: Object): Cursor;
     //http://mongodb.github.io/node-mongodb-native/2.1/api/Cursor.html#setEncoding
-    setEncoding(encoding: string): void;
+    setEncoding(encoding: string): this;
```
Comment (CaselIT) [inline mongodb/index.d.ts]: This

File: mongodb/index.d.ts [hunk 2]
```diff
@@ -1271,7 +1271,7 @@ export interface AggregationCursor extends Readable {
     //http://mongodb.github.io/node-mongodb-native/2.1/api/AggregationCursor.html#rewind
     rewind(): AggregationCursor;
     //http://mongodb.github.io/node-mongodb-native/2.1/api/AggregationCursor.html#setEncoding
-    setEncoding(encoding: string): void;
+    setEncoding(encoding: string): this;
```
Comment (CaselIT) [inline mongodb/index.d.ts]: this

File: mongodb/index.d.ts [hunk 3]
```diff
@@ -1312,7 +1312,7 @@ export interface CommandCursor extends Readable {
     //http://mongodb.github.io/node-mongodb-native/2.1/api/CommandCursor.html#rewind
     rewind(): CommandCursor;
     //http://mongodb.github.io/node-mongodb-native/2.1/api/CommandCursor.html#setEncoding
-    setEncoding(encoding: string): void;
+    setEncoding(encoding: string): this;
```
Comment (CaselIT) [inline mongodb/index.d.ts]: This

**Review by CaselIT — CHANGES_REQUESTED** [review]
The `setEncoding` can be removed from mongodb since it's a parent class method.
Marked the methods below

**Comment by hinell** [comment]
@CaselIT Have done.

**Comment by CaselIT** [comment]
Thanks. That should avoid future problems of the same type.
👍 on the mongodb part

**Comment by hinell** [comment]
@CaselIT Yep. At least on the `mongodb` side. But what about those thousandth others modules? :smirk:

**Comment by CaselIT** [comment]
Well I guess someone will review it.

**Comment by andy-ms** [comment]
Travis is slow because it tests everything depending on `node`, which is a lot! And it looks like it found something.

**Comment by hinell** [comment]
Revision is needed guys.
@andy-ms
@RyanCavanaugh
@minestarks
@paulvanbrenk
@sandersn
@maghis
@zhengbli
@yuit
@zjiekai
@yortus
@xiemaisi
@rogierschouten
@rakatyal
@mhegazy
@iainmcgin
@horiuchi
@HerringtonDarkholme
@jefff
@Flarna
@icopp
@a8775

**Comment by hinell** [comment]
bump

**Comment by rogierschouten** [comment]
:+1: on the node part

**Review by Flarna — APPROVED** [review]
looks fine

**Comment by hinell** [comment]
@rogierschouten 
@Flarna 
Thanks guys.

**Comment by mhegazy** [comment]
@andy-ms any comments?

**Review by ghost — APPROVED** [review]

**Comment by TonyPythoneer** [comment]
@hinell Could you resolve the conflict? thx.

**Comment by hinell** [comment]
Well, seems like this #13968 PR has broken all stuff here. 
@andy-ms Your assistance needed.

**Comment by andy-ms** [comment]
Should be fixed in `master` now.

**Pull request merged**